### ISSUE TITLE
New check: mixed-boolean-operators

### ISSUE THREAD

**Issue by nickdrozd** [pylint-dev/pylint#2339]
Quick, without looking anything up, evaluate the following expressions:

```python
False and False or True
False or False and True
True and False or False
False and True or False
False or False and True
True and True or False
True or False and False
True and True or False and True and False or False
True and False or False and True and True or False
```

Were you able to do them in your head? If so, congratulations, because I certainly couldn't. I can never remember what the rules are for mixed boolean expressions. If it were up to me, I would ban them from the language.

There's an example of this right here in Pylint, in `checkers.typecheck.TypeChecker.visit_assign`:

```python
            for rnode in returns:
                if not (isinstance(rnode.value, astroid.Const)
                        and rnode.value.value is None
                        or rnode.value is None):
                    break
```

Under what circumstances exactly will that loop break? I have no idea, and consequently I cannot modify it.

There should be a check, maybe called `mixed-boolean-operators`, to prevent this.

**Comment by nickdrozd** [comment]
@sushobhit27 , are you interested?

**Comment by sushobhit27** [comment]
@nickdrozd but having checks like these is a very common case IMO.
and if I understood correctly, this should be refactoring check, with suggestion like "club all and/or operation separately using parenthesis".
infact I had suggested to have a check which could have reduced these kind of expresssion to some extent using any/all function, but it was rejected. (https://github.com/PyCQA/pylint/issues/2227)
I would wait for other members inputs on your proposed check before proceeding any further.

**Comment by nickdrozd** [comment]
I still think this is a terrible and confusing pattern, so I'm going to reopen the issue. The checker can be an optional extension.

**Comment by DanielNoord** [comment]
I think this might be more something for an external extension. It feels like this wouldn't serve a very large audience and would increase our maintenance burden quite a bit.
I'm a bit wary of adding non-generally accepted extensions (in lack of a better term) after the recent issue about the `while` extension.

**Comment by Pierre-Sassoulas** [comment]
What's the suggested fix in this case ? Would it be worst performance wise ? It's a very common pattern as was said in previous comment. If not knowing this pattern and not using it means worst performance I don't think we should discourage its use. I feel like this is bordering on something that should be an independent pylint plugin. Let's hear about other maintainers opinions.

**Comment by nickdrozd** [comment]
The suggested fix would be to add parentheses so that the binary operators in a list are uniform -- all `and` or all `or`.

For example, `A and B or C` is better written as `A and (B or C)`. This makes it immediately clear how the expression will be evaluated.

With uniform operators, it becomes safe to swap operands freely. For example, `A and B or C` is equivalent to `A and C or B` but NOT equivalent to `B and A or C`. With sensible parentheses, it would be obvious what's equivalent and what isn't. Without parentheses, swaps cannot be made safely without a good understanding of implicit precedence rules.

I don't see why there should be any performance implications to this.

If this check were enabled, I would expect it to be difficult to address in many cases. This is precisely because it is a confusing pattern! It's hard to understand and therefore it isn't safe to modify without a lot of scrutiny. There is tension between the linter goals of (1) recommending sensible coding practices and (2) not being too annoying.

To reiterate, this is a very bad way to write code. If I saw this in PR review, I would suggest adding parentheses to make the meaning obvious. In "legacy" code, this pattern amounts to a puzzle, and solving the puzzle requires understanding implicit precedence rules in detail.

**Comment by nickdrozd** [comment]
Ruff has this check, with autofix even: https://docs.astral.sh/ruff/rules/parenthesize-chained-operators/

> `and` always binds more tightly than `or` when chaining the two together, but this can be hard to remember (and sometimes surprising). Adding parentheses in these situations can greatly improve code readability, with no change to semantics or performance.
